web/docs/map-data-to-ocsf/README.md

```diff
@@ -0,0 +1,846 @@
+    svc_name: service,
+  },
+  // --- Primary (recommended) ---
+  connection_info: {
+    uid: community_id,
+    direction: direction,
+    direction_id: direction_id,
+    protocol_ver_id: protocol_ver_id,
+    protocol_name: proto,
+    protocol_num: protocol_num,
+  },
+  src_endpoint: {
+    ip: id.orig_h,
+    port: id.orig_p,
+    svc_name: service,
+  },
+  // TODO: we actually could go deeper into the `conn_state` field and
+  // choose a more accurate status. But this would require string
+  // manipulations and/or regex matching, which TQL doesn't have yet.
+  status: "Other",
+  status_code: conn_state,
+  status_id: 99,
+  traffic: {
+    bytes_in: resp_bytes,
+    bytes_out: orig_bytes,
+    packets_in: resp_pkts,
+    packets_out: orig_pkts,
+    total_bytes: orig_bytes + resp_bytes,
+    total_packets: orig_pkts + resp_pkts,
+  },
+  // --- Primary (optional) ---
+  // TODO
+  // - `ja4_fingerprint_list`: once we have some sample logs with JA4
+  //   fingerprints, which requires an additional Zeek package, we should
+  //   populate them here.
+  // - `tls`: if we buffer ssl log for this connection, we could add the
+  //   information in here.
+}
+set @name = "ocsf.network_activity"
+```
+
+</details>
+
+You can imagine that this process is time-consuming when doing it thoroughly, as
+many data sources of dozens of unique event types that each need to be mapped
+individually.
+
+Let's cerebrate real quick what we did:
+
+1. Start at the OCSF event class and map all attributes, partioned by group.
+2. We tried to find counterparts for the attributes in the original event.
+3. Everything that didn't have a counterpart ended in `unmapped`.
+4. In the end, we had a single `this = { .. }` to replace the original event.
+
+We could also use an alternative approach:
+
+1. Consider the original event `this = {unmapped: this}`.
+2. Go to the original event and map each field to its corresponding attribute.
+3. Drop the mapped field from the input.
+4. In the end, `unmapped` will contain the remaining attributes.
+
+### Option B: Map to OCSF
+
+Let's see how the alternate approach looks like. Here's a sample event again:
+
+```json
+{
+  "ts": "2021-11-17T13:45:03.913656",
+  "uid": "C1UE8M1G6ok0h7OrJi",
+  "id": {
+    "orig_h": "35.87.13.125",
+    "orig_p": 8,
+    "resp_h": "198.71.247.91",
+    "resp_p": 0
+  },
+  "proto": "icmp",
+  "service": null,
+  "duration": "58.0us",
+  "orig_bytes": 8,
+  "resp_bytes": 8,
+  "conn_state": "OTH",
+  "local_orig": null,
+  "local_resp": null,
+  "missed_bytes": 0,
+  "history": null,
+  "orig_pkts": 1,
+  "orig_ip_bytes": 36,
+  "resp_pkts": 1,
+  "resp_ip_bytes": 36,
+  "tunnel_parents": null,
+  "community_id": "1:A+WZZOx8yg3UCoNV1IeiSNZUxEk=",
+  "_write_ts": null
+}
+```
+
+As we've already establisehd the mapping from conn.log to OCSF above, let's just
+write the pipeline to illustrate the different approach:
+
+```text title="conn-to-ocsf.tql"
```

Comment:
Here's a slightly different way to structure this:

```
// tql2
read_zeek_tsv
where @name == "zeek.conn"
this = { x: this }
time = x.ts
metadata.uid = x.uid
metadata.logged_time = x._write_ts
src_endpoint.ip = x.id.orig_h
src_endpoint.port = x.id.orig_p
src_endpoint.svc_name = x.service
dst_endpoint.ip = x.id.resp_h
dst_endpoint.port = x.id.resp_p
dst_endpoint.svc_name = x.service
connection_info.uid = x.community_id
if x.local_orig and x.local_resp {
  connection_info.direction = "Lateral"
  connection_info.direction_id = 3
} else if x.local_orig {
  connection_info.direction = "Outbound"
  connection_info.direction_id = 2
} else if x.local_resp {
  connection_info.direction = "Inbound"
  connection_info.direction_id = 1
} else {
  connection_info.direction = "Unknown"
  connection_info.direction_id = 0
}
if x.id.orig_h.is_v6() or x.id.resp_h.is_v6() {
  connection_info.protocol_ver_id = 6
} else {
  connection_info.protocol_ver_id = 4
}
connection_info.protocol_name = x.proto
if x.proto == "tcp" {
  connection_info.protocol_num = 6
} else if x.proto == "udp" {
  connection_info.protocol_num = 17
} else if x.proto == "icmp" {
  connection_info.protocol_num = 1
} else {
  connection_info.protocol_num = -1
}
traffic.bytes_in = x.resp_bytes
traffic.bytes_out = x.orig_bytes
traffic.packets_in = x.resp_pkts
traffic.packets_out = x.orig_pkts
traffic.total_bytes = x.orig_bytes + x.resp_bytes
traffic.total_packets = x.orig_pkts + x.resp_pkts
// drop all the mapped fields
drop (
  x.ts, x.uid, x.id, x.service, x.local_orig, x.local_resp, x.community_id,
  x.proto, x.orig_bytes, x.resp_bytes, x.orig_pkts, x.resp_pkts, x._write_ts,
)
// move the raw event to the back and rename it to unmapped
this = { ...this, unmapped: x }
drop x
@name = "ocsf.network_activity"
```

Note that your example for the second approach (and the above rewritten one) are missing most of the OCSF event's fields.

I would argue that the best approach is in the middle between the two approaches you laid out with the following structure

```
// set input aside
this = {x: this}

// do the mapping, following the order of the OCSF schema
// …

// drop all mapped fields
drop (x.foo, x.bar, x.baz,)
this = {...this, unmapped: x}
drop x

// assign target schema name
@name = "ocsf.xyz"
```